REGRESSION (300337@main): Fix font regression in showPopupMenu

gavin-apple opened the pull request:
#### f50f4cf7ee29f0973ccce366061eccc8932f2f10
<pre>
REGRESSION (300337@main): Fix font regression in showPopupMenu
<a href="https://bugs.webkit.org/show_bug.cgi?id=299483">https://bugs.webkit.org/show_bug.cgi?id=299483</a>
<a href="https://rdar.apple.com/161265062">rdar://161265062</a>

Reviewed by Aditya Keerthi.

Changes CTFont creation in showPopupMenu to support
lookup of system fonts.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm
* Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
(WebKit::WebPopupMenuProxyMac::showPopupMenu):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm:
(TestWebKitAPI::TEST(FontAttributes, SelectTextStyle)):

Canonical link: <a href="https://commits.webkit.org/300763@main">https://commits.webkit.org/300763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e134a931f2632d1dfbc9ed97e58a699841a5a94b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130562 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75933 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1e77a126-6b64-4195-99ce-893b27fdc68b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125660 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52092 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94137 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f3a9858b-fd52-4dd8-b664-8e49aa5d9bb2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35229 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110733 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIAction.ClearTabSpecificActionPropertiesOnNavigation (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74739 "Found 2 new API test failures: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot, TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6e4fad74-3e43-4e41-8c5c-2ab06e32577d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34186 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28893 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74043 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133254 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50735 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38636 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102613 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106952 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102442 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26043 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47785 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26037 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47585 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50589 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56353 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50063 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53409 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51737 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
```diff
@@ -105,23 +105,28 @@
-    RetainPtr<NSFontDescriptor> descriptor = [NSFontDescriptor fontDescriptorWithName:data.postScriptName.createNSString().get() size:scaledFontSize];
-    font = [NSFont fontWithDescriptor:descriptor.get() size:scaledFontSize];
+    RetainPtr<CTFontDescriptorRef> descriptor = adoptCF(CTFontDescriptorCreateWithNameAndSize(data.postScriptName.createCFString().get(), scaledFontSize));
+    RetainPtr<CFArrayRef> matched = adoptCF(CTFontDescriptorCreateMatchingFontDescriptorsWithOptions(descriptor.get(), NULL, kCTFontDescriptorMatchingOptionIncludeHiddenFonts));
```
```suggestion
    RetainPtr descriptor = adoptCF(CTFontDescriptorCreateWithNameAndSize(data.postScriptName.createCFString().get(), scaledFontSize));
    RetainPtr matched = adoptCF(CTFontDescriptorCreateMatchingFontDescriptorsWithOptions(descriptor.get(), NULL, kCTFontDescriptorMatchingOptionIncludeHiddenFonts));
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
```diff
@@ -105,23 +106,28 @@
     // honor the font size, matching other browsers.
     if (!font)
-        font = [NSFont menuFontOfSize:scaledFontSize];
+        font = adoptCF(CTFontCreateUIFontForLanguage(kCTFontUIFontMenuTitle, scaledFontSize, NULL));
```
This effectively hardcodes the implementation of `menuFontOfSize`, which could be redefined by the system.

`NSFont` and `CTFontRef` are toll-free bridged, so I think we should continue using `menuFontOfSize` and bridge cast.

pxlcoder reviewed (commented):
Ideally, the PR/commit title should be prefixed with `REGRESSION (300337@main): `.

pxlcoder reviewed (approved):
Will need to remove `No new tests (OOPS!).` from the commit message in order to merge.

Though I think this change may be testable by swizzling `- (void)popUpMenu:(NSMenu*)menu atLocation:(NSPoint)location width:(CGFloat)width forView:(NSView*)view withSelectedItem:(NSInteger)selectedItem withFont:(NSFont*)font withFlags:(NSPopUpMenuFlags)flags withOptions:(NSDictionary *)options` and asserting on the `font`.

gavin-apple committed:
REGRESSION (300337@main): Fix font regression in showPopupMenu
https://bugs.webkit.org/show_bug.cgi?id=299483
rdar://161265062

Reviewed by Aditya Keerthi.

Changes CTFont creation in showPopupMenu to support
lookup of system fonts.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm
* Source/WebKit/UIProcess/mac/WebPopupMenuProxyMac.mm:
(WebKit::WebPopupMenuProxyMac::showPopupMenu):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm:
(TestWebKitAPI::TEST(FontAttributes, SelectTextStyle)):

Canonical link: https://commits.webkit.org/300763@main
